### PR TITLE
For review

### PR DESCRIPTION
**Pull request by sundapeng**

### CODE REVIEW

File: core/src/main/java/tachyon/Constants.java [hunk 1]
```diff
@@ -4,9 +4,9 @@
  * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
  * "License"); you may not use this file except in compliance with the License. You may obtain a
  * copy of the License at
- *
+ *
```
Comment (sundapeng) [inline core/src/main/java/tachyon/Constants.java]: Remove blank

File: core/src/main/java/tachyon/UnderFileSystemHdfs.java [hunk 2]
```diff
@@ -73,10 +73,11 @@ private UnderFileSystemHdfs(String fsDefaultName, Object conf, TachyonConf tachy
     String glusterfsPrefix = "glusterfs:///";
     tConf.set("fs.defaultFS", fsDefaultName);
     if (fsDefaultName.startsWith(glusterfsPrefix)) {
```
Comment (sundapeng) [inline core/src/main/java/tachyon/UnderFileSystemHdfs.java]: File should be revert

File: core/src/main/java/tachyon/UnderFileSystemHdfs.java [hunk 3]
```diff
@@ -391,4 +393,17 @@ public void setPermission(String path, String posixPerm) throws IOException {
       throw e;
     }
   }
+
+  @Override
+  public void setOwner(String path, String username, String groupname) throws IOException {
+    try {
+      LOG.info("Changing file Owner " + path + " for username: " + username + " groupname: "
```
Comment (sundapeng) [inline core/src/main/java/tachyon/UnderFileSystemHdfs.java]: Log level should be debug

File: core/src/main/java/tachyon/client/TachyonFSCore.java [hunk 4]
```diff
@@ -32,7 +32,7 @@
    * @param path The path of the file
    * @param ufsPath The path of the file in the under file system. If this is empty, the file does
    *        not exist in the under file system yet.
-   * @param blockSizeByte The size of the block in bytes. It is -1 iff ufsPath is non-empty.
```
Comment (sundapeng) [inline core/src/main/java/tachyon/client/TachyonFSCore.java]: should be iff

File: core/src/main/java/tachyon/master/InodeFile.java [hunk 5]
```diff
@@ -91,7 +96,12 @@ static InodeFile loadImage(ImageElement ele) throws IOException {
    * @param creationTimeMs The creation time of the file, in milliseconds
    */
   public InodeFile(String name, int id, int parentId, long blockSizeByte, long creationTimeMs) {
-    super(name, id, parentId, false, creationTimeMs);
```
Comment (sundapeng) [inline core/src/main/java/tachyon/master/InodeFile.java]: "false" should not be used at here, should be "AclUtil".File

File: core/src/main/java/tachyon/master/MasterInfo.java [hunk 6]
```diff
@@ -300,7 +311,13 @@ public MasterInfo(InetSocketAddress address, Journal journal, ExecutorService ex
 
     mRawTables = new RawTables(mTachyonConf);
 
-    mRoot = new InodeFolder("", mInodeCounter.incrementAndGet(), -1, System.currentTimeMillis());
+    mFsOwner = UserGroup.getTachyonLoginUser();
```
Comment (sundapeng) [inline core/src/main/java/tachyon/master/MasterInfo.java]: need let reviewer to know we also do login here

File: core/src/main/java/tachyon/master/MasterInfo.java [hunk 7]
```diff
@@ -2484,4 +2692,71 @@ public void writeImage(ObjectWriter objWriter, DataOutputStream dos) throws IOEx
...
+        if (lastComp || !isDir) {
+          break;
+        }
+        InodeFolder tmpNode = (InodeFolder)curNode;
+        curNode = tmpNode.getChild(pathByNameArr[inodeNum]);
+      }
+      return new InodesInPath(inodes, pathByNameArr);
+    } catch (InvalidPathException e) {
+      throw new RuntimeException("resolve path :" + path.toString() + " error");
+    }
```
Comment (sundapeng) [inline core/src/main/java/tachyon/master/MasterInfo.java]: 2015/5/21 done

File: core/src/main/java/tachyon/Constants.java (hunk 1 above)
Comment (shenguoquan) [inline core/src/main/java/tachyon/Constants.java]: Thanks Dapeng for your remind. I will fix it.

File: core/src/main/java/tachyon/UnderFileSystemHdfs.java (hunk 2 above)
Comment (shenguoquan) [inline core/src/main/java/tachyon/UnderFileSystemHdfs.java]: I will fix it. Thanks

File: core/src/main/java/tachyon/UnderFileSystemHdfs.java (hunk 3 above)
Comment (shenguoquan) [inline core/src/main/java/tachyon/UnderFileSystemHdfs.java]: Thanks dapeng for your advice. I will change it

File: core/src/main/java/tachyon/client/TachyonFSCore.java (hunk 4 above)
Comment (shenguoquan) [inline core/src/main/java/tachyon/client/TachyonFSCore.java]: It's my mistake. I will correct it.

File: core/src/main/java/tachyon/master/TachyonMaster.java [hunk 8]
```diff
@@ -148,6 +150,15 @@ public TachyonMaster(TachyonConf tachyonConf) {
     }
   }
 
+  private TServerSocket createTServerSocket(InetSocketAddress address) throws TTransportException {
+    if (mTachyonConf.getBoolean(Constants.TACHYON_SECURITY_USE_SSL, false)) {
+      // TODO: ssl
+      return null;
```
Comment (sundapeng) [inline core/src/main/java/tachyon/master/TachyonMaster.java]: should throw an unsupported exception.

File: core/src/main/java/tachyon/master/TachyonMaster.java [hunk 9]
```diff
@@ -223,29 +234,44 @@ private void login() throws IOException {
...
+      TTransportFactory tTransportFactory = factory.getAuthTransFactory();
+
+      // create server
+      return new TThreadPoolServer(new TThreadPoolServer.Args(mServerTServerSocket)
+          .maxWorkerThreads(mMaxWorkerThreads).minWorkerThreads(mMinWorkerThreads)
+          .processorFactory(processorFactory).transportFactory(tTransportFactory)
+          .protocolFactory(new TBinaryProtocol.Factory(true, true)));
+    } catch (Exception e) {
+      // TODO: handle it
+      throw new RuntimeException(e);
```
Comment (sundapeng) [inline core/src/main/java/tachyon/master/TachyonMaster.java]: should throw same exception(throws IOException, TTransportException)

File: core/src/main/java/tachyon/security/SecurityUtil.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+package tachyon.security;
+
+import java.io.IOException;
+
+import tachyon.Constants;
+import tachyon.conf.TachyonConf;
+import tachyon.security.authentication.AuthenticationFactory;
+
+public class SecurityUtil {
```
Comment (sundapeng) [inline core/src/main/java/tachyon/security/SecurityUtil.java]: should delete

File: core/src/main/java/tachyon/security/User.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+ * the License.
+ */
+
+package tachyon.security;
+
+import java.security.Principal;
+
+public class User implements Principal {
+  private String mName;
+
```
Comment (sundapeng) [inline core/src/main/java/tachyon/security/User.java]: TODO add attr for kerberos

File: core/src/main/java/tachyon/security/UserGroup.java [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+
+import tachyon.Constants;
+import tachyon.conf.TachyonConf;
+
+public class UserGroup {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
```
Comment (sundapeng) [inline core/src/main/java/tachyon/security/UserGroup.java]: should check the log.

File: core/src/main/java/tachyon/security/UserGroup.java [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+    sGroups = Groups.getUserToGroupsMappingService(conf);
+    UserGroup.sConf = conf;
+  }
+
+  public static synchronized void loginUserFromSubject() throws IOException {
+    ensureInitialized();
+    try {
+      Subject subject = new Subject();
+
+      LoginContext loginContext = newLoginContext("simple", subject,
```
Comment (sundapeng) [inline core/src/main/java/tachyon/security/UserGroup.java]: TODO should support kerberos

File: core/src/main/java/tachyon/security/authentication/AuthenticationFactory.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+      return PlainSaslHelper.getPlainProcessorFactory(service);
+    }
+  }
+
+  public TTransportFactory getAuthTransFactory() throws LoginException {
+    TTransportFactory tTransportFactory;
+
+    if (mAuthTypeStr.equalsIgnoreCase(AuthTypes.KERBEROS.getAuthName())) {
+      //TODO: kerbores
+      throw new LoginException("Unsupported authentication type: " + mAuthTypeStr);
```
Comment (sundapeng) [inline core/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: new UnsupportedException

File: core/src/main/java/tachyon/security/authentication/AuthenticationFactory.java [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    } else if (mAuthTypeStr.equalsIgnoreCase(AuthTypes.NOSASL.getAuthName())) {
+      tTransportFactory =  new TFramedTransport.Factory();
+    } else if (mAuthTypeStr.equalsIgnoreCase(AuthTypes.SIMPLE.getAuthName())) {
+      tTransportFactory = PlainSaslHelper.getPlainTransportFactory(mAuthTypeStr);
+    } else {
+      throw new LoginException("Unsupported authentication type: " + mAuthTypeStr);
```
Comment (sundapeng) [inline core/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Same above

File: core/src/main/java/tachyon/security/authentication/SaslPlainServerProvider.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+  public static class SaslPlainServerFactory implements SaslServerFactory {
+
+    @Override
+    public SaslServer createSaslServer(String mechanism, String protocol, String serverName,
+                                       Map<String, ?> props, CallbackHandler cbh) {
+      if ("PLAIN".equals(mechanism)) {
+        try {
+          return new PlainSaslServer(cbh, protocol);
+        } catch (SaslException e) {
+          /* This is to fulfill the contract of the interface which states that an exception shall
```
Comment (sundapeng) [inline core/src/main/java/tachyon/security/authentication/SaslPlainServerProvider.java]: should delete

File: core/src/main/java/tachyon/web/WebInterfaceBrowseServlet.java [hunk 17]
```diff
@@ -221,6 +222,10 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
           + ie.getMessage());
       getServletContext().getRequestDispatcher("/browse.jsp").forward(request, response);
       return;
+    } catch (AccessControlException ace) {
+      request.setAttribute("fatalError", "Error: Access denied " + ace.getLocalizedMessage());
+      getServletContext().getRequestDispatcher("/memory.jsp").forward(request, response);
```
Comment (sundapeng) [inline core/src/main/java/tachyon/web/WebInterfaceBrowseServlet.java]: check why forward to memory.jsp

File: core/src/main/java/tachyon/web/WebInterfaceDownloadServlet.java [hunk 18]
```diff
@@ -82,6 +83,10 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
     } catch (InvalidPathException ipe) {
       request.setAttribute("invalidPathError", "Error: Invalid Path " + ipe.getLocalizedMessage());
       getServletContext().getRequestDispatcher("/browse.jsp").forward(request, response);
+    } catch (AccessControlException ace) {
+      request.setAttribute("fatalError", "Error: Access denied " + ace.getLocalizedMessage());
+      getServletContext().getRequestDispatcher("/memory.jsp").forward(request, response);
```
Comment (sundapeng) [inline core/src/main/java/tachyon/web/WebInterfaceDownloadServlet.java]: same above

File: core/src/main/java/tachyon/web/WebInterfaceMemoryServlet.java [hunk 19]
```diff
@@ -69,6 +70,10 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
         request.setAttribute("fatalError", "Error: Invalid Path " + ipe.getLocalizedMessage());
         getServletContext().getRequestDispatcher("/memory.jsp").forward(request, response);
         return;
+      } catch (AccessControlException ace) {
+        request.setAttribute("fatalError", "Error: Access denied " + ace.getLocalizedMessage());
+        getServletContext().getRequestDispatcher("/memory.jsp").forward(request, response);
```
Comment (sundapeng) [inline core/src/main/java/tachyon/web/WebInterfaceMemoryServlet.java]: same above

File: core/src/test/java/tachyon/command/TFsShellTest.java [hunk 20]
```diff
@@ -334,7 +340,7 @@ public void locationTest() throws IOException {
   }
 
   @Test
-  public void lsrTest() throws IOException {
+  public void lsrTestest() throws IOException {
```
Comment (sundapeng) [inline core/src/test/java/tachyon/command/TFsShellTest.java]: revert

File: core/src/main/java/tachyon/master/InodeFile.java (hunk 5 above)
Comment (shenguoquan) [inline core/src/main/java/tachyon/master/InodeFile.java]: Good advice. I will fix it

File: core/src/main/java/tachyon/web/WebInterfaceBrowseServlet.java (hunk 17 above)
Comment (shenguoquan) [inline core/src/main/java/tachyon/web/WebInterfaceBrowseServlet.java]: should forward to browse.jsp and show the error information. It's my mistake. I will fix it.

File: core/src/main/java/tachyon/web/WebInterfaceDownloadServlet.java (hunk 18 above)
Comment (shenguoquan) [inline core/src/main/java/tachyon/web/WebInterfaceDownloadServlet.java]: I will fix it

File: core/src/main/java/tachyon/web/WebInterfaceMemoryServlet.java (hunk 19 above)
Comment (shenguoquan) [inline core/src/main/java/tachyon/web/WebInterfaceMemoryServlet.java]: There is correct. It follows the above logic in the WebInterfaceMemoryServlet.java

File: core/src/test/java/tachyon/command/TFsShellTest.java (hunk 20 above)
Comment (shenguoquan) [inline core/src/test/java/tachyon/command/TFsShellTest.java]: ok, I will fix it.

**Pull request closed without merging**